Add Backblaze B2 upload support

Klowner opened the pull request:
This change adds Backblaze B2 bucket upload capability with the addition of an `@uppy/b2-multipart-companion` plugin as well as accompanying controller code for Companion.

The `b2-multipart-companion` plugin introduces one dependency: `js-sha1`

Thanks!

Klowner commented:
:thinking: do individual `packages/` dependencies need to be added to the root `package.json` ?

Klowner commented:
Ah, I need to fix typings and tests as well.

arturi commented:
Hi! This is awesome, thank you! We’ll review as soon as we can 🙏

goto-bus-stop commented on packages/@uppy/companion/src/server/controllers/b2.js:
```diff
@@ -0,0 +1,196 @@
+    const { type } = req.body
+
+    if (typeof fileName !== 'string') {
+      return res.status(500).json({ error: 'b2: filename returned from `getPath` must be a string' })
```
I think 500 is ok, since the getPath() function is defined by the developer, and if it does not return a string it is a server-side issue.

Klowner replied:
ah, yep, at first glance that does look like it should be a `400` like the one below it, but as @goto-bus-stop points out, it's validating the value returned from a user-overrideable method.

Klowner commented on packages/@uppy/companion/src/server/b2.js:
```diff
@@ -0,0 +1,170 @@
+const B2_API_VERSION = 2
+const B2_API_URL = `https://api.backblazeb2.com/b2api/v${B2_API_VERSION}/`
+
+module.exports = class B2Client {
```
I have no particularly strong reason for avoiding the official module, other than the frustrating experience I've had trying to get axios to work in cloud worker environments, and their REST API is very well documented (IMO) and easy to work with directly. But swapping it out should be relatively easy if that would in fact be preferred.

ifedapoolarewaju replied:
I have never tried using Axios in node environment (I'm also not sure what `cloud worker environments` mean here), but if the library is compatible in node environments then I'll be more in favor of using the npm library. This will relieve us of the weight of maintaining the client (for example if `backblazeb2` API gets updated)

Klowner commented on packages/@uppy/companion/src/server/controllers/b2.js:
```diff
@@ -0,0 +1,196 @@
+    const client = req.uppy.b2Client
+    const { fileId } = req.params
+
+    if (typeof fileId !== 'string') {
```
Oh I totally missed that, thank you! The route won't get called unless the param exists, so of course the string will always exist. Good catch.

Klowner commented on packages/@uppy/companion/src/server/Uploader.js:
```diff
@@ -15,6 +15,7 @@ const redis = require('./redis')
 const PROTOCOLS = Object.freeze({
   multipart: 'multipart',
   s3Multipart: 's3-multipart',
+  b2Multipart: 'b2-multipart',
```
This is very likely something I overlooked, could I trouble you to point me towards that section of the S3 stuff? I'm definitely no expert in Uppy's architecture :)

ifedapoolarewaju replied:
it's [here](https://github.com/transloadit/uppy/blob/647bb6497c473a8497a06912a83342fa06d8f6e7/packages/%40uppy/companion/src/server/Uploader.js#L240-L241) and [here](https://github.com/transloadit/uppy/blob/647bb6497c473a8497a06912a83342fa06d8f6e7/packages/%40uppy/companion/src/server/Uploader.js#L497-L553)

Klowner commented on packages/@uppy/companion/src/server/controllers/b2.js:
```diff
@@ -0,0 +1,196 @@
+  }
+
+  const getCachedBucketID = (() => {
+    const cache = Object.create({})
```
That's an outright mistake on my part. I likely meant to use `Object.create(null)` as is used throughout the rest of the codebase. Good catch!

goto-bus-stop commented on packages/@uppy/companion/src/server/b2.js:
```diff
@@ -0,0 +1,170 @@
+   * Request new authorization token from Backblaze.
+   */
+  authorize () {
+    return this._authorizing || (this._authorizing = new Promise((resolve, reject) => {
```
I think this pattern is a bit confusing—could we use one of these instead?
```js
if (this._authorizing) return this._authorizing
this._authorizing = new Promise(...)
return this._authorizing
```
or
```js
if (!this._authorizing) {
  this._authorizing = new Promise(...)
}
return this._authorizing
```

goto-bus-stop commented:
We have some default options for Companion upload targets here:
https://github.com/transloadit/uppy/blob/9abb4e2ba7efd82e12e6b3dc3688f595d238d9d0/packages/%40uppy/companion/src/companion.js#L28-L36

Could we do a similar one for the `getPath()` here for B2?

goto-bus-stop commented:
> :thinking: do individual packages/ dependencies need to be added to the root package.json ?

The root package.json needs a dependency on the plugin inside `packages/`:
```json
"dependencies": {
  "@uppy/backblaze-b2-multipart": "file:packages/@uppy/backblaze-b2-multipart"
}
```
but it doesn't need to list the dependencies of the plugin itself.

goto-bus-stop commented on packages/@uppy/companion/src/standalone/helper.js:
```diff
@@ -57,6 +57,11 @@ const getConfigFromEnv = () => {
           process.env.COMPANION_AWS_USE_ACCELERATE_ENDPOINT === 'true'
+      },
+      b2: {
+        keyId: process.env.COMPANION_B2_KEY_ID,
```
This is used as `applicationKeyId` elsewhere:
```suggestion
        applicationKeyId: process.env.COMPANION_B2_KEY_ID,
```

goto-bus-stop replied:
Hah yeah, renaming all the other things works too :wink:

goto-bus-stop commented on packages/@uppy/companion/src/standalone/helper.js:
```diff
@@ -57,6 +57,11 @@ const getConfigFromEnv = () => {
+      },
+      b2: {
+        keyId: process.env.COMPANION_B2_KEY_ID,
+        key: getSecret('COMPANION_B2_KEY'),
```
similarly:
```suggestion
        applicationKey: getSecret('COMPANION_B2_KEY'),
```

Klowner commented:
I'm pretty certain I addressed most of the feedback items (correctly, I hope) aside from:
 - Swapping out my janky b2 client for `backblaze-b2` package
 - Implement the B2 upload for companion

ifedapoolarewaju commented:
@Klowner all look good so far! Do you think you'll be able to address these last two suggestions?

Klowner commented:
@ifedapoolarewaju thanks! Adding companion uploader support should be no problem, but I'm seeing if I can get automatic 401 response handling added to `backblaze-b2`, if not, I'll have to wrap all the calls anyway :man_shrugging:.

Klowner commented:
`backblaze-b2` also does not yet implement the api calls necessary for resuming large uploads :man_facepalming: 
```
// b2_list_parts
B2.prototype.listParts = notYetImplemented;
// b2_list_unfinished_large_files
B2.prototype.listUnfinishedLargeFiles = notYetImplemented;
```

Klowner commented:
All the companion API calls are switched over to `backblaze-b2`.

I've submitted a PR to implement the absent `listParts()` in `backblaze-b2` which is necessary to resume large file uploads. While I wait for that PR, I'll continue onwards with implementing the companion server to server stuff.

arturi commented:
@Klowner hi! sorry to bother, any updates on this? can we help unblock you somehow?

Klowner commented:
I think I'm getting pretty close to having this done, although I'm running into issues in my testing where files transferred from Dropbox don't seem to have any sensible filename or mimetype metadata being transferred over. I figured I'd inquire as to whether that's a known thing before I dig into that after getting the large file upload fully working.

:wave: :man_dancing:

ifedapoolarewaju commented:
unlike the other providers, the mime-type for dropbox files is [guessed internally by companion](https://github.com/transloadit/uppy/blob/44cc071471c472b129e0a91b5c86f470a737daa0/packages/%40uppy/companion/src/server/provider/dropbox/adapter.js#L24-L30). It tries to detect it from the filename. However, the filename is provider by dropbox, and should be the name that exists from dropbox. But if the filename is busted then the mime-type will be busted because it depends on the filename.

Klowner commented:
I think all that remains is getting the data out from the server-to-server transfers after transfer completion to expose the uploaded fileId, etc. to the client.

Klowner commented:
Also waiting on this PR: https://github.com/yakovkhalinsky/backblaze-b2/pull/92

Klowner commented:
Feel free to review whenever, I don't foresee any remaining significant changes unless I've missed something. :+1:

Klowner commented on packages/@uppy/companion/src/server/Uploader.js:
```diff
@@ -551,6 +568,55 @@ class Uploader {
+      if (error) {
+        this.emitError(error)
+      } else {
+        this.emitSuccess('?', {
```
Ah yes, I should have asked about this. If the destination bucket is public, `b2_download_file_by_id` can be used to get a download URL, but that call is a class B transaction so it comes with a small fee.

Perhaps I expose this as a configuration option? When enabled, returns a valid download URL, making it functionally the same as the s3 response, or `null` if `getDownloadURL` (or some better name) is disabled?

Klowner replied:
Another option may be returning a thunk to make the API call to obtain the download URL

ifedapoolarewaju replied:
> Perhaps I expose this as a configuration option? When enabled, returns a valid download URL, making it functionally the same as the s3 response, or null if getDownloadURL (or some better name) is disabled?

I think we can with this option, yes. It gives the server control over whether or not to allow its usage (since it's costly) 👍

Klowner commented on packages/@uppy/companion/src/server/b2stream.js:
```diff
@@ -0,0 +1,374 @@
+      .catch((err) => {
+        cb(err)
+      })
+      .then((result) => cb(null, result))
```
Not sure what the heck I was thinking with that one.. Looking at it now I'm fairly certain I should have the `.then()` before the `.catch()`.

Klowner commented on packages/@uppy/companion/src/server/b2stream.js:
```diff
@@ -0,0 +1,374 @@
+      this.getBucketId(this.options.bucketName)
+    ]).then(([partSize, bucketId]) => {
+      const fileSize = this.options.fileSize
+      const onUploadProgress = this.createProgressReporter(fileSize)
```
Wholeheartedly agree!

Klowner commented on packages/@uppy/companion/src/server/b2stream.js:
```diff
@@ -0,0 +1,374 @@
+      fdSlicer
+    ]).then(([hash, { slicer }]) =>
+      callWithRetry(attemptPartTransmission, [hash, slicer])
+        .then(({ data }) => {
```
nope, the promise created by then `.then` on line 297 is discarded, so `endpointPool.release`'s result is discarded as well. I can add curly braces around the `endpointPool.release` to make that more explicit.

Klowner commented on packages/@uppy/companion/src/server/Uploader.js:
```diff
@@ -492,18 +502,25 @@ class Uploader {
-   */
-  uploadS3Multipart () {
+  beginTailReadStream () {
     const file = createTailReadStream(this.path, {
```
I'm using the tail read stream in this case to monitor the progress of the file being received so the individual upload parts can begin being hashed as soon as they're received.

Unless your plan is to move away from async download/upload entirely? If that is the case, a lot of this could be simplified I suppose.

ifedapoolarewaju replied:
yes from the note in [this PR](https://github.com/transloadit/uppy/pull/2114), the intention is to move away from parallel upload/download for now, since the usage of the library has seemed [problematic](https://github.com/transloadit/uppy/issues/1457) so far

Klowner commented on packages/@uppy/companion/src/server/b2stream.js:
```diff
@@ -0,0 +1,374 @@
+
+  let handledCounter = 0
+
+  return new Writable({
```
Along with the request to remove the use of tail-stream this is sounding like a significant rewrite at this point and we're already exceeding my cost to benefit ratio for myself.

The "official" b2 client is little more than functions that call axios, and considering it took them a month to get a one line change (and tests and documentation) merged, I don't see them being receptive to actually adding complex logic to their library.

I'm leaning towards just maintaining a personal fork of Uppy and closing this PR.

kvz replied:
That makes sense! The Uppy team’s concern is this is a lot of code quite specific to B2 which is not a majority use case, so we wonder if we can justify the maintenance burden ourselves too. If you could run this as a separate module that Companion uses, that would make things considerably easier for us

Klowner replied:
I hear ya, this is a pretty big PR. I'll explore offloading a majority of the fiddly bits to a glue module.

kisuka commented:
Any progress update on this?

Klowner commented:
@kisuka no

Klowner commented:
Making gradual progress on moving the big moving parts out to an intermediary library. Hopefully after these changes this PR can be significantly shrunk down. :+1:

kvz commented:
Have you seen that B2 offers an s3 compatible api now since a day or two?

I truly hope this doesn’t mean a lot of work goes up in smoke 😯

Sent from mobile, pardon the brevity.

> On 6 May 2020, at 18:23, Mark Riedesel <notifications@github.com> wrote:
> 
> ﻿
> Making gradual progress on moving the big moving parts out to an intermediary library. Hopefully after these changes this PR can be significantly shrunk down. 👍
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

Klowner commented:
@kvz lol, I will be overjoyed if this is a workable solution. Thank you for pointing this out to me, I totally missed that news :)

Klowner commented:
Looks like it's not quite a drop-in replacement. Not sure what parts are causing Uppy to not work, but I'll give it a look.

Klowner commented:
Backblaze people are looking into an issue I found that appears to be related to CORS, I'll post updates as I get 'em.

Klowner commented:
Doesn't look like it'll work until they get CORS support added to their S3 stuff.

Throwing in the towel on this one.

sooxt98 commented:
@Klowner s3 multipart upload is working now with config below

```json
[
  {
    "corsRuleName": "downloadFromAnyOriginWithUpload",
    "allowedOrigins": ["*"],
    "allowedHeaders": [
      "Authorization",
      "x-amz-date",
      "x-amz-content-sha256",
      "content-type"
    ],
    "exposeHeaders": ["ETag"],
    "allowedOperations": [
      "s3_delete",
      "s3_get",
      "s3_head",
      "s3_post",
      "s3_put"
    ],
    "maxAgeSeconds": 3600
  }
]

```

GFoley83 commented:
@Klowner @sooxt98 Did you ever circles back on this one? Do you know if it works now?
